### ISSUE TITLE
Variants (as known in Figma) or States (as known in Adobe XD)

### ISSUE THREAD

**Issue by jcklpe** [penpot/penpot#768]
**Is your feature request related to a problem? Please describe.**  
Yes. Managing components without the ability to create variations/states is a mess. Single biggest issue that prevents me from attempting to use Penpot full time

**Describe the solution you'd like**  
Have states. Pretty straightforward. Don't know what else to say. 

**Describe alternatives you've considered**  
It's possible to make every state/variation it's own component but this is messy. Furthermore states are used in things like Adobe XD to manage things like [microinteractions](https://github.com/penpot/penpot/issues/769) which wouldn't be possible with separate components.

**Comment by jcklpe** [comment]
Here's some videos that demonstrate these features:  
  
[https://www.youtube.com/watch?v=c8ov6IzFE0c](https://www.youtube.com/watch?v=c8ov6IzFE0c)  
  
[https://www.adobe.com/products/xd/learn/prototype/component-states/use-auto-animate-and-component-states.html](https://www.adobe.com/products/xd/learn/prototype/component-states/use-auto-animate-and-component-states.html)

**Comment by elhombretecla** [comment]
Hi @jcklpe! Thanks for your feedback

We already have planned a set of advanced features for components into our road map. As you know, kind of "variants" or "states" functionalities are really complicated and challenging so they deserve the whole team best effort to accomplish them. But we have it in mind!!! Stay tunned :)

**Comment by jcklpe** [comment]
@elhombretecla : any updates on this? This is the only feature really missing IMO for me to start using this in production work.

**Comment by hirunatan** [comment]
Sorry, this is one of the important (and big) features but we still haven't prioritized it high enough.

**Comment by CasparNuel** [comment]
Is there any update about this? Penpot is very close to being my editor-of-choice for wireframing and UX/UI testing. States/variations not being available for components is currently a block for me though.

Expanding, moving or "interacting" components are a nightmare to manage with separate panes. States/Variations being implemented would mean a major leap forward in utility for Penpot.

**Comment by sneko** [comment]
Hi @hirunatan, any news on this?

**Comment by Improov-All** [comment]
We would like to transition to using Penpot, but alas, having multiple artboards to simulate component states is a pain that's blocking most projects.

**Comment by g-clems** [comment]
From what I've seen it's on the 2023 roadmap, so maybe we can expect it before 2024 ? maybe. I hope

https://community.penpot.app/t/faq-what-is-in-penpot-s-roadmap-for-2023

**Comment by jcklpe** [comment]
Let's hope so!

**Comment by sneko** [comment]
@elhombretecla it was in the 2023 roadmap, do you know the exact state of this feature?

Just found the following link that may indicate the task is still pending:
https://tree.taiga.io/project/penpot/us/2088?milestone=262806

**Comment by elhombretecla** [comment]
Hi @sneko :)

We're currently working on a new version of Penpot, Penpot 2.0, which is set to be released soon. This update includes a significant and necessary rework of the component functionality. This overhaul is crucial to address the feature of component variants that you're asking for. You can expect news on this development shortly, and we truly appreciate your interest and comment.

Thank you for staying updated with our progress!

**Comment by sneko** [comment]
@elhombretecla 2.0 is released 😄 ! Do you know where this issue stands in your roadmap now?

EDIT: it seems I can have common properties like shapes between multiple main components, but it still miss the idea of colors... that would make Penpot really useful for design systems.

**Comment by myfunnyandy** [comment]
@sneko this is one of our main focuses after the 2.0 release. The new components version that was included in 2.0 was a necessary step towards Penpot component variations, so we feel that we are closer. 

We are currently working on the definition (you can read some of the advances [here](https://tree.taiga.io/project/penpot/epic/6854)) and hope to start soon the implementation.

**Comment by maxwxyz** [comment]
Any news on this critical feature?

**Comment by enoversum** [comment]
@myfunnyandy Fresh new Penpot user here, and I just realised with my first project that there are no component states, the singular needed functionality for UX design. Do you have any news for this feature?

**Comment by myfunnyandy** [comment]
Hi all, and welcome @enoversum! A good update, we are currently working on what we decided to call Component Variants. You can look at the progress at [its epic](https://tree.taiga.io/project/penpot/epic/6854). Although we don’t have a solid estimate yet, it’s already a key focus for our team.

**Comment by enoversum** [comment]
> Hi all, and welcome [@enoversum](https://github.com/enoversum)! A good update, we are currently working on what we decided to call Component Variants. You can look at the progress at [its epic](https://tree.taiga.io/project/penpot/epic/6854). Although we don’t have a solid estimate yet, it’s already a key focus for our team.

Hi @myfunnyandy , a not-so-funny thank you for your reply and note that you are already working on it. I haven't worked with Penpot enough to know all features in-depth, but I would believe this would mark a milestone in the attractiveness of Penpot once it's finished and shipped into the world. Good luck with building and refining the feature, I'm happy to jump aboard again when it's here 😄 !

**Comment by myfunnyandy** [comment]
@enoversum thanks for the honest answer :)

We hope to come back soon with the super-funny news of the variants release. This isn't the only thing we're cooking up. Just in case you're curious, you can check out our current plans [at this post](https://community.penpot.app/t/not-a-roadmap-recent-current-and-2025-initiatives/7375). The highlight of the next release will very likely be Design Tokens.

**Comment by enoversum** [comment]
@myfunnyandy Aw wow, that sounds lovely, both the new rendering engine and the Design Tokens thing. If I understand correctly, that's some kind of Dev mode that holds together design elements like pre-defined colours, text styles etc. like a library, just like XD and Figma are doing it?

**Comment by myfunnyandy** [comment]
@enoversum sort of, although I wouldn't say "Dev mode" as is a common ground for design and development. Let's say that is a way to define and store design building blocks in a language that can be shared across tools and disciplines. Here's a [wonderful post](https://penpot.app/blog/what-are-design-tokens-a-complete-guide/) explaining it in detail, and here's a [sneak-peek](https://community.penpot.app/t/sneak-peek-tokens-multidimiensional-themes/8205) showing something that you can achieve with them.

**Comment by enoversum** [comment]
@myfunnyandy Yes, and you're absolutely right, it was a bit abbreviated in my expression: Design tokens are the common design decisions in a project. I was thinking more complex and already tying it to the 'Dev mode', as this could and should be reflected in that—instead of seeing ten different colour values which you have to look up all the time to tie it to a colour (variable), you can see something like 'Primary' and just know, this is the primary colour, same with typography and other choices. Fully understood, and it's also a great leap forward, especially together with variations. 

I'm amazed how much work goes into Penpot, and how even more usable it will become soon!

**Comment by jcklpe** [comment]
So excited! Variants are coming to the beta!!!

**Comment by myfunnyandy** [comment]
And you can still join! https://community.penpot.app/t/help-shape-penpot-variants-join-the-beta-test

**Comment by myfunnyandy** [comment]
Hi all, I'm happy to say that Penpot Variants are already live as part of the [2.10 release](https://community.penpot.app/t/penpot-2-10-go-your-own-way/9889) 🚀

**Comment by enoversum** [comment]
> Hi all, I'm happy to say that Penpot Variants are already live as part of the [2.10 release](https://community.penpot.app/t/penpot-2-10-go-your-own-way/9889) 🚀

That is fabulous news! I'll be sure to check them out in the next few days. Happy that you brought this much-needed feature, Penpot is here to stay :) .

**Comment by enoversum** [comment]
@myfunnyandy Hi Andy, just circling back, as I now got a bit of time to spend with Penpot's Component variants. 

Am I right in assuming that there is still no easy way to add a :hover or :focus state to components, like I am used to in Adobe XD? Meaning, I can create a component, create a variant. Then drag the component (default state) onto my layout, e.g. a button, then change the text, switch to the on-layout variant, change text there as well, and I'm good to go? As far as I understand, I can create my component, then a variant, but when I want to have a hover effect, I need to define a mouseenter and mouseleave interaction, which could only link to the original component, so the hover text is broken as well? Still a bit confused, a simpler approach for :hover and :focus states would really be helpful and quite natural for UX designers.

**Comment by jcklpe** [comment]
@enoversum , @myfunnyandy : I actually created a separate feature request for the hover state stuff here: https://github.com/penpot/penpot/issues/769

**Comment by enoversum** [comment]
> [@enoversum](https://github.com/enoversum) , [@myfunnyandy](https://github.com/myfunnyandy) : I actually created a separate feature request for the hover state stuff here: [#769](https://github.com/penpot/penpot/issues/769)

Well, that makes sense! I can't really believe this doesn't sit front and centre for UX design. Given your request is from 2021 and didn't get too much attention yet, I can't really imagine it coming anytime soon 😔.
